### ISSUE TITLE
Repl evaluate doesn't have access to function upvalues

### ISSUE THREAD

**Issue by tmillr** [jbyuki/one-small-step-for-vimkind#39]
I haven't checked, but this may only be specific to callbacks (e.g. autocmd callback, timer callback). It looks like the current impl only uses `debug.getlocal()` while traversing the ***current*** call stack (which may not necessarily contain all of the values the current frame/fn has access to, e.g. its upvalues from its defining context). 

Also, I'm not sure if this is 100% correct regardless, as the outer frame often contains locals which shouldn't technically/normally be accessible from the current frame (although I view this issue as somewhat secondary to the issue concerning a lack of access to upvalues).

### Idea for fix

When evaluating (at least for the repl context, maybe for hover context too?), only collect the locals of the current frame/fn. Then, use `debug.getupvalue()` in a loop to get/collect all of the upvalues of the current frame/fn. 

Precedence/fallback order should be:

```
Matching Local w/ Highest Index || Matching Upvalue || Matching fenv value || nil
```

using the first non-nil value found.

> [!NOTE]
> There can be multiple locals with the same name in a given frame/fn, which is why the one with the highest index wins. Simply using a foreword/increasing loop and collecting the values into a table map should be sufficient, as locals at later/higher indices will naturally overwrite earlier ones.

> [!NOTE]
> Also note how it ends by trying to resolve against the function's env, not _G itself. The env is usually _G, but not always.

There's also a webpage on lua.org describing how to accomplish just this and precisely mimic Lua's usual value retrieval algorithm (i.e. when evaluating/executing some chunk/fn) [here](https://www.lua.org/pil/23.1.2.html).

**Comment by jbyuki** [comment]
Correct, upvalues are not supported by the [evaluate request](https://microsoft.github.io/debug-adapter-protocol//specification.html#Requests_Evaluate) with hover, repl and others. However the  [variables request](https://microsoft.github.io/debug-adapter-protocol//specification.html#Requests_Variables) which I mainly use through `require"dap.ui.widgets".hover()` does support upvalues and if I remember I just copy pasted the code from the link you provided. I will try to add an implementation for upvalues for evaluate requests as well.

**Comment by jbyuki** [comment]
There was another issue actually, the `evaluate` (repl, hover) request was actually not handled properly until now. It was reading all the frames in the call stack during evaluation which it should not. I fixed that so it only reads the current frame. It should also read upvalues in the current frame only now, evaluation of global variables was already handled.

Just to give a bit of deep details if you're interested, it's a bit more subtle than just retrieving the value of variables. Expression such as `1+1` should give `2` in the repl, or even `x+2`. So the line has to be properly executed **with** the local variables context. The `setfenv` comes in handy in that case.

Anyway, I was not sure if you wanted to contribute here but I did it. Feel also free to PR anything in the future. It looks like you know well the inner working of such systems.
